### ISSUE TITLE
Problem with calculating mAP on validation dataset with different difficulty levels

### ISSUE THREAD

**Issue by Berni11** [ultralytics/ultralytics#19607]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I have trained a yolo11n-obb model on the Kitti Dataset by converting the pointclouds into BEV images. The training worked well and the model also detects correctly.

To be able to compare the results with other object detection models, I wanted to validate my yolo model using the 3 difficulties specified on the Kitti Dataset website: Easy, Moderate and Hard
https://www.cvlibs.net/datasets/kitti/eval_object.php?obj_benchmark=bev

For this I have created 3 different validation sets. They contain the same images but have their labels filtered according to the guidelines.

Now when I validate my model using the easy, moderate and hard validation set, I get better mAP on the hard dataset than on the easy dataset. I expected it to be the opposite as the easy dataset has fewer objects to detect.

Here are the validation results from the easy validation set:
`Ultralytics 8.3.86 🚀 Python-3.9.21 torch-2.5.1+cu124 CUDA:0 (Tesla V100-SXM2-32GB, 32494MiB)
YOLO11n-obb summary (fused): 109 layers, 2,654,308 parameters, 0 gradients, 6.6 GFLOPs
val: Scanning /home/user/Dataset/Kitti/BEVdata-obb_val_easy_640/validation/labels.cache... 1481 images, 0 backgrounds, 0 corrupt: 100%|███`

                 `Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 93/93 [00:09<00:00, 10.14it/s]
                   all       1481       2637      0.458      0.779      0.541      0.426
            Pedestrian        571        821      0.608      0.497      0.463      0.299
                   Car        977       1679      0.299      0.993      0.569      0.546
               Cyclist        107        137      0.469      0.847      0.592      0.434
`Speed: 0.3ms preprocess, 1.4ms inference, 0.0ms loss, 1.3ms postprocess per image
Results saved to runs/obb/val125`

Results from the moderate validation set:

`Ultralytics 8.3.86 🚀 Python-3.9.21 torch-2.5.1+cu124 CUDA:0 (Tesla V100-SXM2-32GB, 32494MiB)
val: Scanning /home/user/Dataset/Kitti/BEVdata-obb_val_moderate_640/validation/labels.cache... 1481 images, 13 backgrounds, 0 corrupt: 100`

                 `Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 93/93 [00:09<00:00,  9.73it/s]
                   all       1481       4684      0.696      0.839      0.796      0.608
            Pedestrian        375        779      0.764      0.694      0.739      0.465
                   Car       1235       3695      0.648      0.989      0.859      0.807
               Cyclist        151        210      0.676      0.833      0.791      0.552
`Speed: 0.2ms preprocess, 1.3ms inference, 0.0ms loss, 1.3ms postprocess per image
Results saved to runs/obb/val126`

Results from the hard 

validation set:

`Ultralytics 8.3.86 🚀 Python-3.9.21 torch-2.5.1+cu124 CUDA:0 (Tesla V100-SXM2-32GB, 32494MiB)
val: Scanning /home/user/Dataset/Kitti/BEVdata-obb_val_hard_640/validation/labels.cache... 1481 images, 11 backgrounds, 0 corrupt: 100%|██`

                 `Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 93/93 [00:09<00:00,  9.63it/s]
                   all       1481       6314      0.835      0.838      0.881       0.67
            Pedestrian        396        932      0.814      0.678      0.774      0.479
                   Car       1250       5111      0.879      0.981      0.976      0.907
               Cyclist        193        271      0.811      0.855      0.894      0.624
`Speed: 0.2ms preprocess, 1.3ms inference, 0.0ms loss, 1.3ms postprocess per image`

Image showing the labels from the easy validation set:
![Image](https://github.com/user-attachments/assets/9ee430c2-4686-441b-a7d2-e7774d1b7e92)

Image showing the labels from the moderate validation set:
![Image](https://github.com/user-attachments/assets/1b1b0265-ff77-4030-a302-06e052865800)

Image showing the labels from the hard validation set:
![Image](https://github.com/user-attachments/assets/a10a8902-d48e-499b-b589-c463a3700c0c)

Image with the detections by the model:
![Image](https://github.com/user-attachments/assets/f5d93182-ffc5-4a5c-a335-393cc526d6c9)

Please let me know if someone knows why the mAP values are better on the hard validation set than on the easy validation set and if this can be fixed to get correct mAP values.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Berni11, thank you for your detailed report and for using Ultralytics 🚀! We appreciate your effort in providing the validation results and images. Here's how we can assist you:

We recommend first reviewing our [Docs](https://docs.ultralytics.com/) for guidance, as they contain many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples, which might help clarify some aspects of your validation process.

If this is a 🐛 Bug Report, we kindly ask you to provide a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum-reproducible-example/). This will help us debug the issue effectively. Be sure to include the following:
1. Your validation command(s) and any custom arguments.
2. A small subset of your dataset (if possible), or a clear description of how the validation sets were prepared.
3. Any additional logs or errors not already provided.

If this is a custom training or ❓ Question, you might want to verify the following:
1. Are the validation datasets correctly filtered based on the Kitti Dataset's difficulty levels?
2. Are the validation metrics consistent across different hardware or environments? Sometimes discrepancies can arise from preprocessing or environment configurations.
3. Are you following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)?

To ensure your environment is up-to-date, please upgrade to the latest `ultralytics` version along with its dependencies. Run the following:
```bash
pip install -U ultralytics
```

## Environments

You can validate your model in any of the following verified environments to ensure consistency:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Community and Support

Join the Ultralytics community where it suits you best:
- For real-time assistance, join our [Discord](https://discord.com/invite/ultralytics) 🎧.
- Prefer in-depth discussions? Visit our [Discourse](https://community.ultralytics.com/).
- Share knowledge or ask for advice on our [Subreddit](https://reddit.com/r/Ultralytics).

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If the above badge is green, it means our [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing, and the issue might be specific to your setup.

This is an automated response to help guide you, but rest assured, an Ultralytics engineer will review your issue and provide further assistance soon. Thank you for your patience! 😊

**Comment by Y-T-G** [comment]
The mAP calculation has a bug, which can be fixed by using the suggestion by OP here:

https://github.com/ultralytics/ultralytics/issues/17270

**Comment by Berni11** [comment]
I have tried the mentioned fix but I still get higher mAP on the hard dataset than on the easy dataset:

Easy:

                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 93/93 [00:11<00:00,  8.27it/s]
                   all       1481       2637      0.458      0.779      0.534      0.418
            Pedestrian        571        821      0.608      0.497       0.44       0.28
                   Car        977       1679      0.299      0.993      0.569      0.544
               Cyclist        107        137      0.469      0.847      0.592       0.43

Moderate:

                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 93/93 [00:11<00:00,  8.05it/s]
                   all       1481       4684      0.696      0.839      0.793      0.599
            Pedestrian        375        779      0.764      0.694      0.729      0.452
                   Car       1235       3695      0.648      0.989      0.859      0.799
               Cyclist        151        210      0.676      0.833      0.791      0.545

Hard:

                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 93/93 [00:12<00:00,  7.69it/s]
                   all       1481       6314      0.835      0.838      0.877      0.657
            Pedestrian        396        932      0.814      0.678      0.762      0.463
                   Car       1250       5111      0.879      0.981      0.976      0.895
               Cyclist        193        271      0.811      0.855      0.892      0.614

**Comment by Y-T-G** [comment]
What's your validation code/command?

If you want, you can save the JSON predictions and calculate using COCOEval.

**Comment by Berni11** [comment]
I'm doing validation by doing:

model = YOLO("best.pt")
results = model.val(data="path/to/dataset.yaml")

What's the command to export the predictions to a json file?

**Comment by Y-T-G** [comment]
You can add `save_json=True` to the above command

**Comment by Petros626** [comment]
@Berni11 Hey, I'm trying to achieve exact the same like you. I trained with bev images created from lidar data of the kitti dataset. 

My problem is that I don't know how to evaluate the objects after difficulty level. Can you please help me with that?

**Comment by Petros626** [comment]
@Y-T-G can imagine, why the metrics are getting better, when the difficulty increases? My test setup produces the best metrics for `moderate`, `easy`, `hard`, but easy and moderate should be switched from mA? values.

**Comment by glenn-jocher** [comment]
Looking at your validation results, I notice something interesting: the number of instances increases significantly with difficulty level (Easy: 2637, Moderate: 4684, Hard: 6314). This indicates your datasets are *cumulative* rather than *exclusive*.

In the KITTI evaluation protocol, the "Hard" dataset typically includes all objects (easy, moderate, and hard combined), while "Easy" contains only the subset of objects meeting the easy criteria. This means your model is being evaluated on completely different object distributions across the three sets.

The higher mAP on harder sets makes sense because:

1. The model has more objects to detect, improving statistical robustness
2. The "Hard" set contains all objects from all difficulty levels (including easy ones)
3. Precision and recall calculations are affected by the total number of objects in the dataset

For a more accurate evaluation by difficulty level, you should:

1. Ensure you're using the latest fix for mAP calculation 
2. Consider using KITTI's official evaluation script which properly handles difficulty stratification
3. Try using `save_json=True` and analyze the results with external tools

It's not a bug in your evaluation, but rather a reflection of how the KITTI benchmark stratifies difficulty levels and how object detection metrics work with different distributions of objects.

**Comment by Petros626** [comment]
@glenn-jocher where is it mentioned, that `Hard` includes all objects, `Easy` only the easy samples and `Moderate` only the moderate samples? From their website I understood it, that the objects were seperately evaluated by their difficulty.

**Comment by glenn-jocher** [comment]
@Petros626 The KITTI benchmark doesn't explicitly state the sets are cumulative in that exact wording, but the implementation reveals this approach. Looking at your shared images and @Berni11's data statistics, it's clear that harder difficulty levels contain more objects (Easy: 2,637, Moderate: 4,684, Hard: 6,314).

The official KITTI evaluation method differentiates difficulty by filtering objects during evaluation rather than creating separate validation sets. Each object has an assigned difficulty attribute, and the evaluation includes/excludes objects based on the selected difficulty level.

When evaluating with separate datasets as done here, the increasing performance with difficulty makes mathematical sense: the model has more detection opportunities, and mAP benefits from the larger statistical sample that includes objects from all difficulty categories.

For more accurate difficulty-based evaluation, I recommend:
1. Using the KITTI evaluation code directly: https://github.com/kitti-project/kitti_eval
2. Or implementing their filtering approach within a single dataset evaluation

This would better align with how other implementations report KITTI benchmark results across difficulty levels.

**Comment by Petros626** [comment]
Implementing option 2 was to difficult for me, I tried several approaches, but couldn't get the pipeline working. I try another repo like this: https://github.com/traveller59/kitti-object-eval-python or this https://github.com/tudelft-iv/view-of-delft-dataset/tree/main.

How can I save the output predictions of the model only for gt-labels in ultralytics?

**Comment by glenn-jocher** [comment]
@Petros626 Yes, you should include all detections (both TP and FP) from your YOLO model when using the KITTI evaluation script. The KITTI evaluation methodology is designed to take all predictions and handle the difficulty stratification internally.

The script will:
1. Match your predictions with ground truth
2. Apply the appropriate difficulty criteria to each object
3. Calculate metrics separately for each difficulty level

This approach ensures consistent evaluation across different models and implementations. Make sure your prediction format matches what the KITTI evaluation script expects - typically this includes 3D bounding boxes, confidence scores, and class labels.

**Comment by Petros626** [comment]
Allright thanks for me the whole issue is clarified.

**Comment by glenn-jocher** [comment]
@Petros626 Glad the issue is clarified. To summarize for anyone who might encounter similar situations with KITTI evaluation:

1. The counterintuitive results (better mAP on "hard" datasets) occur because the difficulty levels in KITTI are cumulative, with the hard set containing more objects overall

2. For proper difficulty-based evaluation using YOLO models:
   - Export predictions with `save_json=True` and `save_txt=True`
   - Use the official KITTI evaluation scripts which correctly handle difficulty stratification
   - Include all detections (both TP and FP) in your evaluation

This approach will give you metrics that align with the standard KITTI benchmark reporting methodology across difficulty levels.

**Comment by Petros626** [comment]
Finally solved it using an official kitti evaluation code (https://github.com/open-mmlab/OpenPCDet/tree/master/pcdet/datasets/kitti/kitti_object_eval_python, https://github.com/traveller59/kitti-object-eval-python):

![Image](https://github.com/user-attachments/assets/1ef78942-f433-4d10-8349-56a68798d66e)

You can check my repos, which I used to implement a custom kitti bev dataset, training the yolo detector and evaluate it.

**Comment by glenn-jocher** [comment]
@Petros626 Excellent work! Your results perfectly demonstrate the expected difficulty progression (Easy > Moderate > Hard in terms of performance) that aligns with proper KITTI evaluation methodology.

This is a great example of why using official evaluation scripts is crucial for specialized benchmarks like KITTI - the built-in YOLO validation wasn't designed to handle KITTI's specific difficulty stratification approach, but the official tools handle it correctly.

Thanks for sharing your solution and repo links - this will be valuable for others working with KITTI BEV detection tasks using YOLO models.
